shared/configs.ts

```diff
@@ -13,9 +13,27 @@ function minimatch(file: string, pattern: string) {
   return m.match(file)
 }
 
-export function getMatchedGlobs(file: string, glob: (string | string[])[]) {
-  const globs = (Array.isArray(glob) ? glob : [glob]).flat()
-  return globs.filter(glob => minimatch(file, glob)).flat()
+export function getMatchedGlobs(file: string, globs: (string | string[])[]) {
+  const flatGlobs = (Array.isArray(globs) ? globs : [globs]).flat()
+  let unmatched: string[] = []
+
+  flatGlobs.forEach((glob, i) => {
+    if (minimatch(file, glob)) {
+      if (glob.startsWith('!'))
+        unmatched.push(glob)
+    }
+    else {
+      if (glob.startsWith('!')) {
+        const unignoreMatched = getMatchedGlobs(file, flatGlobs.slice(0, i))
+        unmatched = unmatched.concat(unignoreMatched.length > 0 ? unignoreMatched : [], glob)
+      }
+      else {
+        unmatched.push(glob)
+      }
+    }
+  })
+
+  return flatGlobs.filter(glob => !unmatched.includes(glob))
```

Comment:
I guess the order of ignore and unignore might matter how it work. For the must accurate result, we might need to vendor ESLint's walker here: https://github.com/eslint/eslint/blob/5db226f4da9ad7d53a4505a90290b68d4036c082/lib/eslint/eslint-helpers.js#L216

Reply:
@antfu I just updated the PR

I took a different approach, cos i realised that `@eslint/config-array` has a very convenient `isFileIgnored` [here](https://github.com/eslint/rewrite/tree/main/packages/config-array#determining-ignored-paths), we can use this to decide if a file is ignored, and that should fix the issue entirely.

I don't think we need to vendor the eslint-helper since it's not exported and there are a long chain of dependencies in eslint's implementations.